IdentityCore/src/cache/ios/MSIDKeychainTokenCache.m

```diff
@@ -624,5 +625,88 @@ - (BOOL)clearWithContext:(id<MSIDRequestContext>)context
     return YES;
 }
 
+- (BOOL)saveAppMetadata:(MSIDAppMetadataCacheItem *)item
+                key:(MSIDCacheKey *)key
+         serializer:(id<MSIDAppMetadataItemSerializer>)serializer
+            context:(id<MSIDRequestContext>)context
+              error:(NSError **)error
+{
+    assert(item);
+    assert(serializer);
+    
+    NSData *itemData = [serializer serializeAppMetadataCacheItem:item];
+    
+    if (!itemData)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Failed to serialize app metadata item.", nil, nil, nil, context.correlationId, nil);
+        }
+        MSID_LOG_ERROR(context, @"Failed to serialize app metadata item.");
+        return NO;
+    }
+    
+    MSID_LOG_INFO_PII(context, @"Save keychain item, item info %@", item);
+    
+    return [self saveData:itemData
+                      key:key
+                  context:context
+                    error:error];
+}
+
+- (MSIDAppMetadataCacheItem *)appMetadataWithKey:(MSIDCacheKey *)key
+                serializer:(id<MSIDAppMetadataItemSerializer>)serializer
+                   context:(id<MSIDRequestContext>)context
+                     error:(NSError **)error;
+{
+    MSID_LOG_INFO(context, @"Get app metadata, key info (account: %@ service: %@ generic: %@ type: %@)", _PII_NULLIFY(key.account), key.service, _PII_NULLIFY(key.generic), key.type);
+    MSID_LOG_INFO_PII(context, @"Get app metadata, key info (account: %@ service: %@ generic: %@ type: %@)", key.account, key.service, key.generic, key.type);
+    
+    NSMutableDictionary *query = [self.defaultKeychainQuery mutableCopy];
+    if (key.service)
+    {
+        [query setObject:key.service forKey:(id)kSecAttrService];
+    }
+    if (key.account)
+    {
+        [query setObject:key.account forKey:(id)kSecAttrAccount];
+    }
+    if (key.generic)
+    {
+        [query setObject:key.generic forKey:(id)kSecAttrGeneric];
+    }
+    if (key.type != nil)
+    {
+        [query setObject:key.type forKey:(id)kSecAttrType];
```

Comment:
can you reuse itemsWithKey:context:error: here to avoid code duplication? The code looks very similar to me.

Reply:
itemsWithKey:context:error returns an array. But metadata will always be one right. I can reuse that api but then I will have to get the first item and do additional processing to cast the first item to MSIDAppMetadataCacheKey. I think we can leave it this way.